diff --git a/AmberTools/src/sander/mdread2.F90 b/AmberTools/src/sander/mdread2.F90
index 29e2207e..082cc327 100644
--- a/AmberTools/src/sander/mdread2.F90
+++ b/AmberTools/src/sander/mdread2.F90
@@ -550,6 +550,11 @@
          write(6, '(/a)') 'Bussi temperature regulation:'
          write(6, '(5x,3(a,f10.5))') 'temp0   =', temp0, &
                ', tempi   =', tempi, ', tautp   =', tautp
+      else if( ntt == 12) then
+         write(6,'(/a)') 'Stochastic Isokinetic Nose-Hoover RESPA (SINR) integration (Middle Scheme):'
+         write(6,'(5x,3(a,f10.5))') 'temp0   =',temp0, &
+               ', tempi   =',tempi,', gamma_ln=', gamma_ln
+         write(6,'(5x,a,i10,a,f12.5,/)') 'nkija   =',nkija,', sinrtau  =',sinrtau
       end if
 
       if( ntp /= 0 ) then
@@ -2010,7 +2015,7 @@
       DELAYED_ERROR
    end if
 
-   if (ntt < 0 .or. ntt > 11) then                                      ! APJ
+   if (ntt < 0 .or. ntt > 12) then                                      ! APJ
       write(6,'(/2x,a,i3,a)') 'NTT (',ntt,') must be between 0 and 11.' ! APJ
       DELAYED_ERROR
    end if
@@ -2018,9 +2023,9 @@
       write(6, '(/2x,a,f6.2,a)') 'TAUTP (',tautp,') < DT (step size)'
       DELAYED_ERROR
    end if
-   if( ntt < 3 .or. ntt > 10) then  ! APJ
+   if( ntt < 3 .or. (ntt > 10 .and. ntt /= 12)) then  ! APJ
       if( gamma_ln > 0.d0 ) then
-         write(6,'(a)') 'ntt must be 3 to 10 if gamma_ln > 0' ! APJ
+         write(6,'(a)') 'ntt must be 3 to 10 or 12 if gamma_ln > 0' ! APJ
          DELAYED_ERROR
       end if
    end if
@@ -2067,7 +2072,7 @@
       end if
    end if
 
-   if (ntt == 10) then
+   if (ntt == 10 .or. ntt == 12) then
       write(6,*) ""
       if(gamma_ln <= 0.d0) then
          write(6,'(a)') 'gamma_ln must be > 0 when ntt = 10'
@@ -2085,8 +2090,8 @@
          write(6,'(a)') 'tempi must be <= temp0 when ntt = 10'
          DELAYED_ERROR
       end if
-      if(sinrtau < 0.5) then
-         write(6,'(a)') 'sinrtau must be >= 0.5 when ntt = 10'
+      if(sinrtau <= 0.d0) then
+         write(6,'(a)') 'sinrtau must be > 0 when ntt = 10'
          DELAYED_ERROR
       end if
   end if
diff --git a/AmberTools/src/sander/runmd.F90 b/AmberTools/src/sander/runmd.F90
index 19a993fd..a90bc78e 100644
--- a/AmberTools/src/sander/runmd.F90
+++ b/AmberTools/src/sander/runmd.F90
@@ -330,7 +330,7 @@ subroutine runmd(xx, ix, ih, ipairs, x, winv, amass, f, v, vold, xr, xc, &
   ! Variables to control a Langevin dynamics simulation
   logical is_langevin
   _REAL_ gammai, c_implic, c_explic, c_ave, sdfac, ekins0
-  _REAL_ dtx, dtxinv, dt5, factt, ekin0, ekinp0, dtcp, dttp
+  _REAL_ dtx, dtxinv, dt5, factt, ekin0, ekinp0, dtcp, dttp, sinrtaux
   _REAL_ rndf, rndfs, rndfp, boltz2, pconv, tempsu
   _REAL_ xcm(3), acm(3), ocm(3), vcm(3), ekcm, ekrot
   _REAL_ emtmd
@@ -1643,18 +1643,19 @@ subroutine runmd(xx, ix, ih, ipairs, x, winv, amass, f, v, vold, xr, xc, &
 
    ! }}}
   end if
-  if (ntt == 10) then
+  if (ntt == 10 .or. ntt == 12) then
      !   initialize the SINR integrator: {{{
 
+    sinrtaux = sinrtau * 20.455d0
 #ifdef MPI
-    call sinr_init(natom, nkija, dtx, boltz2, temp0, gammai, sinrtau, &
+    call sinr_init(natom, nkija, dtx, boltz2, temp0, gammai, sinrtaux, &
                    sinrdata, commsander)
 #else
-    call sinr_init(natom, nkija, dtx, boltz2, temp0, gammai, sinrtau, &
+    call sinr_init(natom, nkija, dtx, boltz2, temp0, gammai, sinrtaux, &
                    sinrdata)
 #endif
     if (irest == 1) then
-      call sinr_read_vels(v, sinrtau, sinrdata)
+      call sinr_read_vels(v, sinrtaux, sinrdata)
     else
       call init_sinr_vels(v, amass, sinrdata)
     endif
@@ -1664,12 +1665,14 @@ subroutine runmd(xx, ix, ih, ipairs, x, winv, amass, f, v, vold, xr, xc, &
       call sinr_mpi_init(v,numtasks,sinrdata)
     end if
 #endif
-    call force(xx, ix, ih, ipairs, x, f, ener, ener%vir, &
-               xx(l96), xx(l97), xx(l98), xx(l99), qsetup, &
-               do_list_update,nstep)
-    call iLndt(v, amass, istart, iend, sinrdata)
-    call iLvdt(v, f, amass, istart, iend, sinrdata)
-    call iLudt(x, v, istart, iend, sinrdata)
+    if (ntt == 10) then
+      call force(xx, ix, ih, ipairs, x, f, ener, ener%vir, &
+                 xx(l96), xx(l97), xx(l98), xx(l99), qsetup, &
+                 do_list_update,nstep)
+      call iLndt(v, amass, istart, iend, sinrdata)
+      call iLvdt(v, f, amass, istart, iend, sinrdata)
+      call iLudt(x, v, istart, iend, sinrdata)
+    end if
     ! }}}
   endif
 
@@ -2689,18 +2692,28 @@ subroutine runmd(xx, ix, ih, ipairs, x, winv, amass, f, v, vold, xr, xc, &
     ! }}}
 
 !------------------------------------------------------------------------------
-  else if (ntt == 10) then
+  else if (ntt == 10 .or. ntt == 12) then
   ! Stochastic Isokinetic Nose-Hoover RESPA (SINR) integrator) {{{
-    call iLvdt(v, f, amass, istart, iend, sinrdata)
-    call iLndt(v, amass, istart, iend, sinrdata)
-    call iLndt(v, amass, istart, iend, sinrdata)
-    call iLvdt(v, f, amass, istart, iend, sinrdata)
-    call iLudt(x, v, istart, iend, sinrdata)
+    if (ntt == 10) then
+      call iLvdt(v, f, amass, istart, iend, sinrdata)
+      call iLndt(v, amass, istart, iend, sinrdata)
+      call iLndt(v, amass, istart, iend, sinrdata)
+      call iLvdt(v, f, amass, istart, iend, sinrdata)
+      call iLudt(x, v, istart, iend, sinrdata)
+    else
+      call iLvdt(v, f, amass, istart, iend, sinrdata)
+      call iLvdt(v, f, amass, istart, iend, sinrdata)
+      call iLqdt(x, v, istart, iend, sinrdata)
+      call iLndt(v, amass, istart, iend, sinrdata)
+      call iLodt(istart, iend, sinrdata)
+      call iLndt(v, amass, istart, iend, sinrdata)
+      call iLqdt(x, v, istart, iend, sinrdata)
+    end if
     if (mod(nstep+1, ntpr*nrespa) == 0) then
       call sinr_temp(v, amass, istart3, iend3, sinrdata)
     end if
     if (mod(nstep+1, ntwr) == 0 .or. nstep+1 == nstlim) then
-      call sinr_write_vels(v, nstep, istart3, iend3, sinrtau, sinrdata)
+      call sinr_write_vels(v, nstep, istart3, iend3, sinrtaux, sinrdata)
     end if
     ! }}}
 
@@ -3072,7 +3085,7 @@ subroutine runmd(xx, ix, ih, ipairs, x, winv, amass, f, v, vold, xr, xc, &
       f(i3) = x(i3)
       x(i3) = aa*(x(i3)*aa + v(i3)*poly*dtx)
     end do
-  else if (ntt == 10) then
+  else if (ntt == 10 .or. ntt == 12) then
     do i3 = istart3, iend3
       f(i3) = x(i3)
     end do
@@ -4925,7 +4938,7 @@ subroutine runmd(xx, ix, ih, ipairs, x, winv, amass, f, v, vold, xr, xc, &
   ! End Replica Exchange MD post-dynamics work }}}
 
   ! Stochastic Isokinetic Nose-Hoover RESPA (SINR) integrator clean-up
-  if (ntt==10) call sinr_cleanup(sinrdata)
+  if (ntt==10 .or. ntt==12) call sinr_cleanup(sinrdata)
 
   ! Print averages {{{
 #ifdef MPI
diff --git a/AmberTools/src/sander/sinr_t.F90 b/AmberTools/src/sander/sinr_t.F90
index a4dc2f48..89a4bb09 100644
--- a/AmberTools/src/sander/sinr_t.F90
+++ b/AmberTools/src/sander/sinr_t.F90
@@ -33,7 +33,8 @@ module sinr_t
 
   end type sinr
 
-public sinr, sinr_init, init_sinr_vels, iLndt, iLvdt, iLudt, sinr_cleanup, &
+public sinr, sinr_init, init_sinr_vels, sinr_cleanup, &
+       iLndt, iLvdt, iLudt, iLqdt, iLodt, &
        sinr_temp, sinr_write_vels, sinr_read_vels
 
 #ifdef MPI
@@ -108,7 +109,7 @@ contains
      sd%nsys = 3
      sd%nres = 2
 
-     sd%wj(1) = (1.0d0/(2.0d0 - 2.0d0**(1.0d0/3.0d0)))*sd%dt/dble(sd%nres)
+     sd%wj(1) = (1.0d0/(2.0d0 - 2.0d0**(1.0d0/3.0d0)))*sd%dt2/dble(sd%nres)
      sd%wj(3) = sd%wj(1)
      sd%wj(2) = -2.0d0**(1.0d0/3.0d0)*sd%wj(1)
 
@@ -118,7 +119,7 @@ contains
      if(sd%mpirank==0) then
 #endif
         write(6,*) ""
-        write(6,'(A74)') "| Using Stochastic Isokinetic Nose-Hoover RESPA (SINR) integrator (ntt=10)"
+        write(6,'(A77)') "| Using Stochastic Isokinetic Nose-Hoover RESPA (SINR) integrator (ntt=10/12)"
         write(6,'(A74)') "| ------------------------------------------------------------------------"
         write(6,*) ""
         write(6,'(A71)') "| NOTE: Only the coordinates are canonical while the velocites are NOT."
@@ -213,6 +214,9 @@ contains
      call mpi_bcast(sd%v1,sd%L*3*sd%natom,MPI_DOUBLE_PRECISION, &
                     0,sd%mpicomm,ierr)
 
+     call mpi_bcast(sd%v2,sd%L*3*sd%natom,MPI_DOUBLE_PRECISION, &
+                    0,sd%mpicomm,ierr)
+
      sd%nproc = ntask
 
    end subroutine sinr_mpi_init
@@ -257,7 +261,7 @@ contains
                ! Outer translation operator LN,2 eq. 66
 
                  do k=1,sd%L
-                    sd%v2(k,ind+jj) = sd%v2(k,ind+jj) + 0.25d0*sd%wj(i)*(sd%Q1*(sd%v1(k,ind+jj)**2) - sd%kbT)/sd%Q2
+                    sd%v2(k,ind+jj) = sd%v2(k,ind+jj) + 0.5d0*sd%wj(i)*(sd%Q1*(sd%v1(k,ind+jj)**2) - sd%kbT)/sd%Q2
                  enddo
 
                ! Inner translation operator LN,1 eq. 66, solutions with eq. 70
@@ -265,7 +269,7 @@ contains
                  q1v1sq = 0.0d0
 
                  do k=1,sd%L
-                    v2kw(k) = exp(-1.0d0*sd%v2(k,ind+jj)*sd%wj(i))
+                    v2kw(k) = exp(-2.0d0*sd%v2(k,ind+jj)*sd%wj(i))
                     q1v1sq = q1v1sq + sd%Q1*sd%v1(k,ind+jj)*sd%v1(k,ind+jj)*v2kw(k)
                  enddo
 
@@ -274,13 +278,13 @@ contains
                  v(ind+jj) = v(ind+jj)*Htnres
 
                  do k=1,sd%L
-                    sd%v1(k,ind+jj) = sd%v1(k,ind+jj)*Htnres*v2kw(k)
+                    sd%v1(k,ind+jj) = sd%v1(k,ind+jj)*Htnres*sqrt(v2kw(k))
                  enddo
 
                ! LN,2 again
 
                  do k=1,sd%L
-                    sd%v2(k,ind+jj) = sd%v2(k,ind+jj) + 0.25d0*sd%wj(i)*(sd%Q1*(sd%v1(k,ind+jj)**2) - sd%kbT)/sd%Q2
+                    sd%v2(k,ind+jj) = sd%v2(k,ind+jj) + 0.5d0*sd%wj(i)*(sd%Q1*(sd%v1(k,ind+jj)**2) - sd%kbT)/sd%Q2
                  enddo
 
               enddo
@@ -298,30 +302,25 @@ contains
      integer istart,iend
      integer ind,i,j,k
      double precision :: v(3*sd%natom),f(3*sd%natom),m(sd%natom)
-     double precision :: sinhbt,coshbt,s,sdot,a,b,sb,sbt
+     double precision :: sinhbt,coshbt,fs,sdot,sbt,p,mlbeta,smlbeta
 
      ind = 3*(istart-1)
 
      do i=istart,iend !sd%natom
         do j=1,3
 
-           a = f(ind+j)*v(ind+j)/sd%lbeta
-           b = (f(ind+j)**2)/(m(i)*(sd%lbeta))
+           p = m(i)*v(ind+j)
+           mlbeta = m(i)*(sd%lbeta)
+           smlbeta = sqrt(mlbeta)
 
-           sb = sqrt(b)
-           sbt = sb*(sd%dt2)
+           sbt = (f(ind+j)/smlbeta)*(sd%dt2)
+           sinhbt = sinh(sbt)
+           coshbt = cosh(sbt)
 
-           if(sbt < sd%errtol) then
-              s = ((((b*a/24.0d0)*sd%dt2 + b/6.0d0)*sd%dt2 + 0.5d0*a)*sd%dt2 + 1.0d0)*sd%dt2
-              sdot = (((b*a/6.0d0)*sd%dt2 + 0.5d0*b)*sd%dt2 + a)*sd%dt2 + 1.0d0
-           else
-              sinhbt = sinh(sbt)
-              coshbt = cosh(sbt)
-              s = (1.0d0/sb)*sinhbt + (a/b)*(coshbt - 1.0d0)
-              sdot = (a/sb)*sinhbt + coshbt
-           endif
+           fs = smlbeta*sinhbt + p*(coshbt - 1.0d0)
+           sdot = (p/smlbeta)*sinhbt + coshbt
 
-           v(ind+j) = (v(ind+j) + (f(ind+j)/m(i))*s)/sdot
+           v(ind+j) = (v(ind+j) + (fs/m(i)))/sdot
 
            ! Update thermostats
 
@@ -359,6 +358,44 @@ contains
      enddo
    end subroutine iLudt
 
+   subroutine iLqdt(q,v,istart,iend,sd)
+     implicit none
+     type(sinr), intent(inout) :: sd
+     integer istart,iend
+     integer ind,i,j
+     double precision :: q(3*sd%natom),v(3*sd%natom)
+
+     ind = 3*(istart-1)
+
+     do i=istart,iend
+        do j=1,3
+           q(ind+j) = q(ind+j) + sd%dt2*v(ind+j)
+        enddo
+        ind=ind+3
+     enddo
+   end subroutine iLqdt
+
+   subroutine iLodt(istart,iend,sd)
+     use random
+     implicit none
+     type(sinr), intent(inout) :: sd
+     integer istart,iend
+     integer ind,i,j,k
+     double precision :: rnum
+
+     ind = 3*(istart-1)
+
+     do i=istart,iend
+        do j=1,3
+           do k=1,sd%L
+              call gauss(0.0d0,1.0d0,rnum)
+              sd%v2(k,ind+j) = sd%v2(k,ind+j)*sd%egt + rnum*sd%sigsqe2gt
+           enddo
+        enddo
+        ind=ind+3
+     enddo
+   end subroutine iLodt
+
    subroutine sinr_temp(v,m,istart,iend,sd)
      implicit none
 #ifdef MPI
